wireless_modules/middle_module/MQTT_Client_class.py

```diff
@@ -0,0 +1,100 @@
+    from umqtt.simple import MQTTClient
+except FileNotFoundError:
+    print('Error in importing library')
+
+
+class Client:
+    def __init__(self, client_id, mqtt_broker):
+        """
+        Initialises the MQTT Client
+        :param client_id: The unique client id used to initiate an MQTTClient class
+        :param mqtt_broker: A string holding domain name or IP address of the broker to connect to, to send and receive
+                            data.
+        """
+        self.client = MQTTClient(client_id, mqtt_broker)
+        self.mqtt_broker = mqtt_broker
+
+    def connect_and_subscribe(self, topics_to_subscribe, callback_func):
+        """
+        Connects to the MQTT broker and subscribes to topic in 'topics_to_subscribe'
+        :param topics_to_subscribe: An array of topics to subscribe to.
+                                    Eah element must be a string or byte literal (the latter is preferred)
+        :param callback_func: The function to be called whenever a message from the subscribed topics is received.
+        :return: True if a connection to the Broker is successfully established, otherwise False
+        """
+
+        # Provide the function to run whenever a message is received from the subscribed topics
+        self.client.set_callback(callback_func)
+        try:
+            # Connect to MQTT broker
+            self.client.connect()
+            print('Connected to %s MQTT broker' % (self.mqtt_broker))
+        except OSError:
+            return False
+
+        # Subscribe to each topic
+        for topic in topics_to_subscribe:
+            self.client.subscribe(topic)
+            print('Subscribed to %s topic' % (topic))
+
+        return True
+
+    def _to_bytes_literal(self, data):
+        """
+        Converts 'data' into a form MQTT can read
+        :param data: A string of data to convert to bytes literal
+        :return: The bytes literal version of the 'data'
+        """
+        str_data = str(data)
+        return str.encode(str_data)
+
+    def mqtt_pub(self, data, topic):
+        """
+        This functions takes care of all of the formatting and publishes 'data' on the given 'topic'. It also checks for
+        any pending message, but this should not be relied upon solely - use the check_for_message() method externally
+        when an incoming message is to be checked for/read.
+        :param data: A string of data to be sent
+        :param topic: A string representing the topic to send 'data' to.
+        :return: Nothing
+        """
+        try:
+            # convert the data to bytes literal so that MQTT can read it
+            msg = self._to_bytes_literal(data)
+
+            # Publish the data on the given topic
+            self.client.publish(topic, msg)
+
+            # Check for any pending messages on any of the subscribed topics
+            self.check_for_message()
+
+            # print("MQTT data sent: %s on %s through %s" % (msg, topic, self.mqtt_broker))
```

Comment:
Don't really like commented out code. Either keep it or remove it. In this case, it might be good to have it?

Reply:
True, fair point. But commenting is something I do solely for myself - I like it when comments can just spell out in English what most complex code bits do so I can debug and understand it much better (but from your feedback I can see I did over comment in a lot of places lol). And I guess for large projects it would make the files much longer and make the code more cloudy, so I'll try to keep it down in the future!